### DISCUSSION TITLE
How to set data as read only?

### DISCUSSION THREAD

**Discussion by marrale** [vestrel00/contacts-android#301] [Q&A] [answered]
One more question, is in some way possible to add the "ContactsContract.Data.IS_READ_ONLY" flag on add/update with the actual release?
Or is something that could be implemented?
Thanks Alessandro

**Comment by vestrel00** [comment]
I have not played around with that field. I'll take a look at it this week to see what it's all about =)

**Reply by marrale** [reply]
Thanks, here a snipped from the source code how I use it:
`val contentProviderOperationBuilder = ContentProviderOperation.newInsert(ContactsContract.Data.CONTENT_URI)
    .withValueBackReference(
        ContactsContract.Data.RAW_CONTACT_ID,
        rawId
    )
    .withValue(ContactsContract.Data.IS_READ_ONLY, 1) // Readonly == true
    .withValue(
        ContactsContract.Data.MIMETYPE,
        ContactsContract.CommonDataKinds.StructuredName.CONTENT_ITEM_TYPE
    )`

**Comment by vestrel00** [comment]
I do have a question. The title of this discussion is about setting Contacts to be read-only, yet the ContactsContract constant you are referencing is specific to Data table rows. 

**Are you interested in making data table rows read-only or also RawContact table rows too?**

I remember in a [different discussion](https://github.com/vestrel00/contacts-android/discussions/224#discussioncomment-2950542) I was unable to read/write the value of `ContactsContract.RawContacts.RAW_CONTACT_IS_READ_ONLY`. Making a RawContact read-only not only makes all of its data rows read-only but it probably also makes it so that the RawContact cannot be modified or deleted except by sync adapters.

Since you are working on sync adapters, you have higher privileges than regular applications. This library is tailored for application usage, not sync adapters. So, I don't want to add too much sync-adapter-only functionality into the library (at least for now). However, if it can also be used by regular applications in some way, I will be inclined to add support for it into this library =) 

Anyways, I will have a deeper look into read-only RawContacts and Data rows this week.

**Reply by marrale** [reply]
I have not used a sync adapter until now. I don't need the sync adapter for my purposes either.
The READ_ONLY flag does not need a "sync adapter". 
If you set the READ_ONLY flag, the contact can not be updated. But you can delete the contact without problems, despite the "READ_ONLY" flag.
Instead of an update, I delete the contact and add it again.
As long as you can set the "READ_ONLY" flag via the API, that's enough for me. This way I can just use the API and don't have to write anything custom.

**Reply by vestrel00** [reply]
Here's the official documentation on `ContactsContract.DataColumns.IS_READ_ONLY`,

![Screenshot 2023-08-30 at 8 19 52 AM](https://github.com/vestrel00/contacts-android/assets/4332753/74bfb1fa-f3b1-4f86-a40b-450f4ae97909)

According to the docs, only sync adapters are allowed to update or delete **DATA rows** that are read-only. Sure, we may be able to insert data rows as read-only. I'll try that. I will also look into expanding all CRUD operations this API provides to allow users to set  `CALLER_IS_SYNCADAPTER` to true (default false) in order to support update and delete operations (as a sync adapter).

No promises though. I will look more into all of the different read-only flags in `ContactsContract` this week to get a more holistic understanding of how it works.

- `ContactsContract.DataColumns.IS_READ_ONLY`
- `ContactsContract.RawContactsColumns.RAW_CONTACT_IS_READ_ONLY`
- `ContactsContract.GroupsColumns.GROUP_IS_READ_ONLY`
    - _This library is actually already using this one._

**Reply by marrale** [reply]
Thanks, I'm interest about the outcome after your testes with the different read only flag. Can be that I wrongly use the field, but it looks working also whiteout sync adapter.
_>  I will also look into expanding all CRUD operations this API provides to allow users to set CALLER_IS_SYNCADAPTER to true (default false) in order to support update and delete operations (as a sync adapter)_
This can be a great improvement to.

**Reply by vestrel00** [reply]
I spent a little bit of time digging. Here's what I have so far.

-----------
Attempting to read the `ContactsContract.RawContactsColumns.RAW_CONTACT_IS_READ_ONLY` column throws this exception...

```
java.lang.IllegalArgumentException: Invalid column raw_contact_is_read_only
```

I tried on a Google Pixel 4 API 33 emulator and on my physical Google Pixel 6 Pro API. I don't know what's going on there. I did not find anything either when searching the internet...

So, I'll have to skip `RAW_CONTACT_IS_READ_ONLY` for now. I don't think it's possible to incorporate into this API at the moment. It's also not what you are asking for anyways 😄 

-----------
As for the other read-only flag in ContactsContract, `ContactsContract.GroupsColumns.GROUP_IS_READ_ONLY`, the Groups APIs and entities provided in this library already uses it and works fine AFAIK.

-----------
Next, I'll focus on investigating `ContactsContract.DataColumns.IS_READ_ONLY`, which is what you are asking about. This is specific to **DATA** rows of RawContacts. It indicates whether a **data row** (e.g. email, phone, name, etc) is read-only or not. It does NOT indicate whether a Contact or RawContact is read-only. Just making sure we are on the same page 🤔

**Reply by vestrel00** [reply]
Hmmm.... I'm also getting an exception when attempting to read `ContactsContract.DataColumns.IS_READ_ONLY`...

```
java.lang.IllegalArgumentException: Invalid column is_read_only
```

@marrale, can you try logging it? Do you also get an exception? You can use the following code to log it,

```kotlin
fun Context.logDataTable() {
    val cursor = contentResolver.query(
        ContactsContract.Data.CONTENT_URI,
        arrayOf(
            ContactsContract.Data._ID,
            ContactsContract.Data.RAW_CONTACT_ID,
            ContactsContract.Data.CONTACT_ID,
            ContactsContract.Data.IS_READ_ONLY
        ),
        null,
        null,
        null
    )

    cursor ?: return

    while (cursor.moveToNext()) {
        // Use getString instead of getLong, getInt, etc so that the value could be null.
        val id = cursor.getString(0)
        val rawContactId = cursor.getString(1)
        val contactId = cursor.getString(2)
        val readOnly = cursor.getString(3)

        log("Data id: $id, rawContactId: $rawContactId, contactId: $contactId, readOnly: $readOnly")
    }

    cursor.close()
}
```

If reading the read-only value is not possible, I'm less inclined to add it to the APIs...

---------

Next, I will take a look at the effects of inserting data rows with `ContactsContract.DataColumns.IS_READ_ONLY` set to true. Hopefully, it won't throw an exception 🤣

**Reply by vestrel00** [reply]
Okay, so... Here's what I found... We can set the the value of `ContactsContract.DataColumns.IS_READ_ONLY` on insert. It will actually result in a read-only data row. 

Take this for example... Using the sample app in this repo with some local code changes, I created a RawContact named "Bob". It has two data rows. One is read-only and the other is not.

- name: "Bob", is_read_only: "1"
- email: "bob@example.com", is_read_only: "0"

Here is what happens when attempting to change the name and email values using Contacts apps...

https://github.com/vestrel00/contacts-android/assets/4332753/b588e507-64bf-4d40-adec-eb13e289531a

https://github.com/vestrel00/contacts-android/assets/4332753/c8574251-71eb-4459-af3f-c39f4eb0d8ce

https://github.com/vestrel00/contacts-android/assets/4332753/88ff5051-ae6d-4ffb-9d71-706b7822dd80

**All of the above contacts app shows that the update succeeded. The email did get updated BUT the name did not because it is read-only.** Note that even if only the name is changed, the apps still show updated succeeded. This means that we cannot really check if the update truly applied the changes (unless we perform a query and compare before-and-after). This is just a limitation we will have to live with.

If `ContactsContract.CALLER_IS_SYNCADAPTER` is set to "1" (true), read-only data rows are able to be updated....

https://github.com/vestrel00/contacts-android/assets/4332753/7b0553e6-9a7e-4089-be73-a35fb2cb08b9

Of course, it will be up to the Account's sync adapter to update the backend in order to persist changes across devices.

-------

As for `ContactsContract.RawContactsColumns.RAW_CONTACT_IS_READ_ONLY`, it did not behave the same way as `ContactsContract.DataColumns.IS_READ_ONLY`. The one for RawContacts did not have any behavioral effects AFAIK. So, I'll completely ignore that for now.

**Reply by vestrel00** [reply, marked as answer]
With all that said, I created several issues that should support your use case. I don't necessarily agree that these things should be added to this library. However, I can see value in it since it can help devs that are trying to do the same or similar things as you.

- #306 
- #307
- #308
- #309

I just have to make sure these new additions are properly documented so that average users of this library do not use them unnecessarily or incorrectly.

I will include all of the above issues in the next release, [0.3.1](https://github.com/vestrel00/contacts-android/discussions/288). I'm busy with work and life so I may not be able to make the release until later next week 😄

**Reply by marrale** [reply]
@vestrel00 thanks for taking your time for this exhaustive testing and explanation. And thanks to add the "ContactsContract.DataColumns.IS_READ_ONLY" to your library, this will really help me for my purpose and I'm able to fully use your library.

About the "ContactsContract.RawContactsColumns.RAW_CONTACT_IS_READ_ONLY" I guess is totally related to Sync adapter. I never try it, if I will give it a try in the future. As I said before, the "ContactsContract.DataColumns.IS_READ_ONLY" fulfil my purpose.

About the release, don't hurry, I will be very happy if you will be able to release it end of next week, but also if it is later on, no problem. Thanks a lot.

**Reply by vestrel00** [reply]
@marrale, I completed all of the issues and am ready to release the next version ([0.3.1](https://github.com/vestrel00/contacts-android/discussions/288)) next week! In addition to the [`SOURCE_ID` discussions we had](https://github.com/vestrel00/contacts-android/discussions/299#discussioncomment-6867208), here are some more new things that you will be able to do...

Insert new RawContacts with data set to read-only,

```kotlin
contactsApi
        .insert()
        .rawContact {
            addEmail {
                isReadOnly = true
                address = "123@abc.com"
                type = EmailEntity.Type.WORK
            }
            setName {
                isReadOnly = true
                givenName = "Abc"
            }
            setNickname {
                isReadOnly = true
                name = "Alphabet"
            }
            setNote {
                isReadOnly = true
                note = "The english alphabet"
            }
            ...
        }
        .commit()
```

I also added an extension function so that you can set the read-only property of all data of the RawContact in one call,

```kotlin
contactsApi
        .insert()
        .rawContact {
            addEmail {
                address = "123@abc.com"
                type = EmailEntity.Type.WORK
            }
            setName {
                givenName = "Abc"
            }
            setNickname {
                name = "Alphabet"
            }
            setNote {
                note = "The english alphabet"
            }
            ...
            // Make sure to call this after you've set/added all RawContact data.
            setDataAsReadOnly(true)
        }
        .commit()
```

[As discussed](https://github.com/vestrel00/contacts-android/discussions/301#discussioncomment-6879925), I am unable to include the `ContactsContract.DataColumns.IS_READ_ONLY` column in the projections array of queries because it throws an exception. This means that `ExistingDataEntity` (e.g. `Name`, `Email`, etc) does NOT have a `readOnly` property. However, I did implement extension functions that will allow you to check if any existing data is read-only...

```kotlin
val readOnly: Boolean? = rawContact.name?.isReadOnly(contactsApi)
val readOnlyMap: Map<Long, Boolean> = rawContact.dataList().isReadOnlyMap(contactsApi)
```

Additionally, you are now able to create instances of the `contacts.core.Contacts` API with `ContactsContract.CALLER_IS_SYNCADAPTER` set to true, which will be applied to all CRUD operations performed using that Contacts instance...

```kotlin
val contactsApi = Contacts(context, callerIsSyncAdapter = true)
```

---------------------

@marrale, please let me know if these changes suffice. If you need anything else, please let me know so that I can include it in the next release (if it is not too much effort) 🙏

**Reply by marrale** [reply]
@vestrel00  thanks a lot for your effort, really appreciated. It looks like your improvments includes everything (and more) needed for my purpose. Especially I love this extension function "setDataAsReadOnly(true)". I'm looking forward to the library release of next week. BR Alessandro

**Reply by vestrel00** [reply]
🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥 
This is now release in [0.3.1](https://github.com/vestrel00/contacts-android/releases/tag/0.3.1)!
🔥 🔥 🔥 🔥 🔥 🔥 🔥 🔥

**Reply by marrale** [reply]
Thanks @vestrel00 for adding this new useful feature!

**Reply by marrale** [reply]
@vestrel00 is possible to do a small "cosmetic improvement": Add the "isReadOnly" as property in "NewRawContact"?
Actual Code is:
```
fun toNewRawContact(): NewRawContact {
    val newRawContact = NewRawContact(
        sourceId = id,
        name = NewName(
            givenName = firstName,
            middleName = middleName,
            familyName = lastName
        ),
        account = Account(ACCOUNT_NAME, ACCOUNT_TYPE)
    )
    newRawContact.setDataAsReadOnly(true)
    return newRawContact
}
```
Whished code:
```
fun toNewRawContact(): NewRawContact {
    val newRawContact = NewRawContact(
        sourceId = id,
        isReadOnly = true,
        name = NewName(
            givenName = firstName,
            middleName = middleName,
            familyName = lastName
        ),
        account = Account(ACCOUNT_NAME, ACCOUNT_TYPE)
    )
    return newRawContact
}
```

**Reply by vestrel00** [reply]
Unfortunately, I don't think that's a good idea. Remember that the RawContact itself is not read only. Only the data is. I know it is "cosmetic" but users might mistake it for `ContactsContract.RawContactsColumns.RAW_CONTACT_IS_READ_ONLY`, which we are currently not using.

IMO, it's better to be explicit and state "data is read only" instead of "raw contact is read only".

**Reply by marrale** [reply]
You are right, and something like this?
```
fun toNewRawContact(): NewRawContact {
    val newRawContact = NewRawContact(
        sourceId = id,
        dataIsReadOnly = true,
        name = NewName(
            givenName = firstName,
            middleName = middleName,
            familyName = lastName
        ),
        account = Account(ACCOUNT_NAME, ACCOUNT_TYPE)
    )
    return newRawContact
}
```

**Reply by vestrel00** [reply]
~~Sure! That's doable. I will need to change the contract of the existing extension function since I will be making this into a property. Actually, it'll be a breaking change. I will delete the `setDataAsReadOnly` extension function and replace it with a property. I'll create an issue for this later today and push out another alpha tag for you~~

**Reply by vestrel00** [reply]
Actually, sorry. I think it's best to not make it a property of `NewRawContact` because it will increase code complexity and/or negatively affect performance a little bit. 

The current extension function `setDataAsReadOnly` loops through all of the already-added data and set the read only property of each data. Each data (e.g. email, name, etc) has its own read only flag to allow users of the API to pick and choose which data is read only and not. 

Supporting this behavior as a property of `NewRawContact` complicates things because the API has no control of how the individual data is created and set. We can set this property at the `NewRawContact` level but users can set the read only flag of some data to true and some false afterwards. This will lead to inconsistency between the top-level `NewRawContact` flag and the individual data flags.

We could increase code complexity and maybe sacrifice a bit of performance to make sure that the `NewRawContact` property is in sync with all of the data property at all times... BUT IMO it's not worth the cosmetic upgrade (even though I agree that it would look nicer). Sorry 🙏

**Reply by marrale** [reply]
That's fine, I agree with you, to many work and complexity for only "cosmetic". By posting the question I was not aware of this complexity.
